Failed to compile with autofree

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.10 dbc4071, press to see full `v doctor` output</summary>

|V full version      |V 0.4.10 55d004e310048c6f3205e06ce0b41a94cc4e95af.dbc4071
|:-------------------|:-------------------
|OS                  |windows, Microsoft Windows 11 Pro 26100 64-bit
|Processor           |32 cpus, 64bit, little endian, 13th Gen Intel(R) Core(TM) i9-13900K  
|Memory              |45.66GB/63.77GB
|                    |
|V executable        |D:\v\v.exe
|V last modified time|2025-06-03 04:59:18
|                    |
|V home dir          |OK, value: D:\v
|VMODULES            |OK, value: C:\Users\kalll\.vmodules
|VTMP                |OK, value: C:\Users\kalll\AppData\Local\Temp\v_0
|Current working dir |OK, value: C:\Users\kalll\repos\v\ki_test
|                    |
|Git version         |git version 2.47.1.windows.1
|V git status        |weekly.2025.23-6-g8c573cf3
|.git/config present |true
|                    |
|cc version          |N/A
|gcc version         |gcc (x86_64-posix-seh, Built by strawberryperl.com project) 8.3.0
|clang version       |N/A
|msvc version        |N/A
|tcc version         |tcc version 0.9.27 (x86_64 Windows)
|tcc git status      |N/A
|emcc version        |N/A
|glibc version       |N/A

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg -autofree src/main.v && C:\Users\kalll\repos\v\ki_test\src\main.exe`
```v
module main

import encoding.binary
import math

fn main() {
	value := Vector3D{
		x: 1.0
		y: 2.0
		z: 3.0
		v: "bob"
	}
	mut buf := []u8{len: 0, cap: 12}
	serialize_to(value, mut buf)!
	println(buf)
}

type Primitive = f64 | f32 | rune | i32 | u32 | i16 | u16 | i8 | u8 | bool

struct Vector3D {
	x f32
	y f32
	z f32
        v string
}

fn serialize_to[T](val T, mut output []u8)! {
	$for v in Primitive.variants {
		$if v.typ is T {
			output << binary.encode_binary(val, binary.EncodeConfig {buffer_len: int(sizeof[T]()), big_endian: false})!
			return
		}
	}
	$if T is string {
		if val.len > math.maxof[u16]() {
			return error('String too long to serialize')
		}
		bytes := val.bytes()
		output << binary.encode_binary(u16(bytes.len), binary.EncodeConfig {buffer_len: int(sizeof[u16]()), big_endian: false})!
		output << bytes
		return
	}
	$for field in T.fields {
		serialize_to(val.$(field.name), mut output)!
	}
}
```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: C:/Users/kalll/AppData/Local/Temp/v_0/main.01JWYA6H9NAP9QAWFBVAJ5QE0J.tmp.c:552: warning: WINVER redefined
cc: C:/Users/kalll/AppData/Local/Temp/v_0/main.01JWYA6H9NAP9QAWFBVAJ5QE0J.tmp.c:7562: error: '{' expected (got ";")
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C:\Users\kalll\AppData\Local\Temp\v_0\ki_test.01JWYASV1Y9X57DMHZXAK8MMKZ.tmp.c: In function 'main__serialize_to_T_main__Vector3D':
C:\Users\kalll\AppData\Local\Temp\v_0\ki_test.01JWYASV1Y9X57DMHZXAK8MMKZ.tmp.c:7538:44: error: invalid initializer
 string _arg_expr_main_serialize_to_1_867 = val.x;// new af2 pre/* inserted before */
                                            ^~~
C:\Users\kalll\AppData\Local\Temp\v_0\ki_test.01JWYASV1Y9X57DMHZXAK8MMKZ.tmp.c:7539:68: error: incompatible type for argument 1 of 'main__serialize_to_T_f32'
       _result_void _t12 = main__serialize_to_T_f32(/*autofree arg*/_arg_expr_main_serialize_to_1_867, output);
                                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\Users\kalll\AppData\Local\Temp\v_0\ki_test.01JWYASV1Y9X57DMHZXAK8MMKZ.tmp.c:1855:43: note: expected 'f32' {aka 'float'} but argument is of type 'string' {aka 'struct string'}
 _result_void main__serialize_to_T_f32(f32 val, Array_u8* output);
                                       ~~~~^~~
C:\Users\kalll\AppData\Local\Temp\v_0\ki_test.01JWYASV1Y9X57DMHZXAK8MMKZ.tmp.c:7567:44: error: invalid initializer
 string _arg_expr_main_serialize_to_1_867 = val.y;// new af2 pre/* inserted before */
                                            ^~~
C:\Users\kalll\AppData\Local\Temp\v_0\ki_test.01JWYASV1Y9X57DMHZXAK8MMKZ.tmp.c:7568:68: error: incompatible type for argument 1 of 'main__serialize_to_T_f32'
       _result_void _t14 = main__serialize_to_T_f32(/*autofree arg*/_arg_expr_main_serialize_to_1_867, output);
                                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\Users\kalll\AppData\Local\Temp\v_0\ki_test.01JWYASV1Y9X57DMHZXAK8MMKZ.tmp.c:1855:43: note: expected 'f32' {aka 'float'} but argument is of type 'string' {aka 'struct string'}
 _result_void main__serialize_to_T_f32(f32 val, Array_u8* output);
                                       ~~~~^~~
C:\Users\kalll\AppData\Local\Temp\v_0\ki_test.01JWYASV1Y9X57DMHZXAK8MMKZ.tmp.c:7596:44: error: invalid initializer
 string _arg_expr_main_serialize_to_1_867 = val.z;// new af2 pre/* inserted before */
                                            ^~~
C:\Users\kalll\AppData\Local\Temp\v_0\ki_test.01JWYASV1Y9X57DMHZXAK8MMKZ.tmp.c:7597:68: error: incompatible type for argument 1 of 'main__serialize_to_T_f32'
       _result_void _t16 = main__serialize_to_T_f32(/*autofree arg*/_arg_expr_main_serialize_to_1_867, output);
                                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\Users\kalll\AppData\Local\Temp\v_0\ki_test.01JWYASV1Y9X57DMHZXAK8MMKZ.tmp.c:1855:43: note: expected 'f32' {aka 'float'} but argument is of type 'string' {aka 'struct string'}
 _result_void main__serialize_to_T_f32(f32 val, Array_u8* output);
                                       ~~~~^~~
```
I must note this error is only happening with autofree. and seems to only occur when its a mix of type in the fields.

**What did you expect to see?**

successful compilation.



> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -731,7 +731,7 @@ fn (mut c Checker) call_expr(mut node ast.CallExpr) ast.Type {
 			if arg.typ != ast.string_type {
 				continue
 			}
-			if arg.expr in [ast.Ident, ast.StringLiteral, ast.SelectorExpr]
+			if arg.expr in [ast.Ident, ast.StringLiteral, ast.SelectorExpr, ast.ComptimeSelector]
 				|| (arg.expr is ast.CallExpr && arg.expr.or_block.kind != .absent) {
 				// Simple expressions like variables, string literals, selector expressions
 				// (`x.field`) can't result in allocations and don't need to be assigned to

--- a/vlib/v/slow_tests/valgrind/comptime_selector.v
+++ b/vlib/v/slow_tests/valgrind/comptime_selector.v
@@ -1,0 +1,52 @@
+module main
+
+import encoding.binary
+import math
+
+fn main() {
+	value := Vector3D{
+		x: 1.0
+		y: 2.0
+		z: 3.0
+		v: 'bob'
+	}
+	mut buf := []u8{len: 0, cap: 12}
+	serialize_to(value, mut buf)!
+	println(buf)
+}
+
+type Primitive = f64 | f32 | rune | i32 | u32 | i16 | u16 | i8 | u8 | bool
+
+struct Vector3D {
+	x f32
+	y f32
+	z f32
+	v string
+}
+
+fn serialize_to[T](val T, mut output []u8) ! {
+	$for v in Primitive.variants {
+		$if v.typ is T {
+			output << binary.encode_binary(val, binary.EncodeConfig{
+				buffer_len: int(sizeof[T]())
+				big_endian: false
+			})!
+			return
+		}
+	}
+	$if T is string {
+		if val.len > math.maxof[u16]() {
+			return error('String too long to serialize')
+		}
+		bytes := val.bytes()
+		output << binary.encode_binary(u16(bytes.len), binary.EncodeConfig{
+			buffer_len: int(sizeof[u16]())
+			big_endian: false
+		})!
+		output << bytes
+		return
+	}
+	$for field in T.fields {
+		serialize_to(val.$(field.name), mut output)!
+	}
+}

--- a/vlib/v/slow_tests/valgrind/valgrind_test.v
+++ b/vlib/v/slow_tests/valgrind/valgrind_test.v
@@ -32,6 +32,7 @@ const skip_valgrind_files = [
 	'vlib/v/slow_tests/valgrind/option_simple.v',
 	'vlib/v/slow_tests/valgrind/string_plus_string_plus.v',
 	'vlib/v/slow_tests/valgrind/import_x_json2.v',
+	'vlib/v/slow_tests/valgrind/comptime_selector.v',
 ]
 
 fn vprintln(s string) {

